diff --git a/meson.build b/meson.build
index d73a8fd4f..22366dad1 100644
--- a/meson.build
+++ b/meson.build
@@ -166,8 +166,6 @@ check_funcs = [
 ]
 
 check_thread_flags = [
-  [['-D_REENTRANT'], ['-lpthread']],
-  [['-pthread'], []],
   [['-D_REENTRANT'], [], {'real': false}],
 ]
 
@@ -794,8 +792,6 @@ else
   cairoboilerplate_dep = dependency('', required: false)
 endif
 
-subdir('util')
-
 if not get_option('tests').disabled() and feature_conf.get('CAIRO_HAS_PNG_FUNCTIONS', 0) == 1
   subdir('test')
   subdir('perf')
